### ISSUE TITLE
Pods need to pre-declare service links iff they want the environment variables created

### ISSUE THREAD

**Issue by bgrant0607** [kubernetes/kubernetes#1768]
Forked from #1107 and #386.

Now seems like a good time to decide whether we want to require/encourage/allow pods to declare services they depend upon. Internally, we've often wished we had such a mechanism. 

Not only would pre-declaration reduce accidental/lazy coupling, but it would also improve scalability by reducing the number of iptables rules that must be created. Pre-declaration would also be compatible with Docker's approach to links. If we supported service aliasing in these declarations, that would facilitate dependency injection for tests and a wide variety of deployment adaptation scenarios, which seems like a compelling alternative to custom environment variables, command-line flags, dynamic configuration services, and so on.

However, it at least needs possible to opt out of static declaration and/or enforcement of service dependencies, such as in the case that dependent services are registered dynamically -- think of a proxy, load balancer, web browser, monitoring service, or naming/discovery service running in a container. 

We'd also definitely need to support more flavors of services (cardinal services, headless services, master election, sharding, ...) in order for most clients to be able to utilize the pre-declaration mechanism.

Something else to consider is how to address dependencies pulled in by client libraries, though perhaps it's not unreasonable to require client libraries to be transparent regarding which services they access.

/cc @thockin @smarterclayton

**Comment by pmorie** [comment]
@bgrant0607 

The following use-case requires predeclaration:

1.  As a user I want to be able to customize how a service is consumed by allowing the name and
    form of environment variables to be transformed prior to use by a pod so that images which are
    not designed to work with a specific named service can be used with kubernetes

Predeclaration would also play a role in the cross-namespace use-case:

1.  As a user I want a pod to be able to consume the metadata of a service in a different
    namespace (example: namespace A contains a database service which I want to use from namespace B)

I'll also note that there is a use-case where you want to opt-out of predeclaration and get info / firewall rules for every service in a pod's namespace.  This adds a wrinkle because it divides pods into two types, those that follow the normal rules and those that don't, which would definitely impact scheduling in order to keep an opted-out pod from wrecking port allocation on a host where all other pods predeclare.  I think @erictune's suggestion of defaulting to fully connected within a namespace and requiring predeclaration of cross namespace dependencies is a good middle ground.

@bgrant0607 @erictune @smarterclayton As a next step, how about a PR to explore the predeclaration mechanism in a vacuum.  I would suggest that `PodSpec` be changed as follows:

    type PodSpec struct {
      // other fields omitted
      ServiceLinks ServiceLinkList
    }

    type ServiceLinkList struct {
      TypeMeta
      ListMeta
      Items     []ServiceLink
    }

    type ServiceLink struct {
      TypeMeta
      ObjectMeta
      Name          string
      Namespace  string
      // Better perhaps as an ObjectReference?
   }

The first iteration could change the `BoundPodFactory` or move the service env functionality to the Kubelet, the latter of which seems like the direction the tide is going.

**Comment by pmorie** [comment]
@lavalamp ^

**Comment by lavalamp** [comment]
Moving BoundPodFactory stuff to kubelet is good and necessary, but @erictune may be working on that already.

If we switch to DNS and deprecate env vars completely, does that eliminate the need for predeclaration? That sounds much easier...

**Comment by pmorie** [comment]
@lavalamp I don't think switching to DNS and deprecating env vars fully eliminates the need for predeclaration.  There there's still the problem of iptables rules on the node.

**Comment by smarterclayton** [comment]
You still have to know what DNS you're looking forward to, and software that runs on different namespaces or clusters won't have the same DNS.

You need something injected into the container that lets legacy software react to the cluster topology.  Predeclaration for adaption is a key thing, there's lots of software out there that doesn't know anything about X_SERVICE_HOST at all.

Also, that same software has to work outside of a cluster as well - on a local dev box how would you point your app to your db (except by using env or mutating a file on disk)?

**Comment by smarterclayton** [comment]
> On Jan 9, 2015, at 12:46 PM, Paul Morie <notifications@github.com> wrote:
> 
> @bgrant0607
> 
> The following use-case requires predeclaration:
> 
> As a user I want to be able to customize how a service is consumed by allowing the name and form of environment variables to be transformed prior to use by a pod so that images which are not designed to work with a specific named service can be used with kubernetes
> Predeclaration would also play a role in the cross-namespace use-case:
> 
> As a user I want a pod to be able to consume the metadata of a service in a different namespace (example: namespace A contains a database service which I want to use from namespace B)
> I'll also note that there is a use-case where you want to opt-out of predeclaration and get info / firewall rules for every service in a pod's namespace. This adds a wrinkle because it divides pods into two types, those that follow the normal rules and those that don't, which would definitely impact scheduling in order to keep an opted-out pod from wrecking port allocation on a host where all other pods predeclare. I think @erictune's suggestion of defaulting to fully connected within a namespace and requiring predeclaration of cross namespace dependencies is a good middle ground.
> 
> @bgrant0607 @erictune @smarterclayton As a next step, how about a PR to explore the predeclaration mechanism in a vacuum. I would suggest that PodSpec be changed as follows:
> 
> type PodSpec struct {
>   // other fields omitted
>   ServiceLinks ServiceLinkList
> }
> 
> type ServiceLinkList struct {
>   TypeMeta
>   ListMeta
>   Items     []ServiceLink
> }
> 
> type ServiceLink struct {
>   TypeMeta
>   ObjectMeta
>   Name          string
>   Namespace  string
>   // Better perhaps as an ObjectReference?
> }
> 
I had been envisioning this to do adaptation, so mutating *how* the service shows up in the pod.  I think that would make the use case a bit more concrete and practical.

**Comment by pmorie** [comment]
@smarterclayton We can roll adaptation into the POC, I will think through a design and propose a model here

**Comment by erictune** [comment]
@pmorie can you give a more concrete example of a system with legacy software that needs service links?

**Comment by pmorie** [comment]
@erictune As a more detailed example, say I have an image that depends on a specially formatted environment variable with a URL for a service.  As an example format, take:

    DOCKER_URL=http://$DOCKER_HOST/$DOCKER_PORT

This use case is to be able to adapt to these special requirements an image may have without changing the image.

It's definitely the case that there's a pretty sizable subset of these cases that can be addressed by performing the translation via the shell and either setting containers' commands to set the variables or wrapping an image with another image containing a script.

It's arguable that for those cases, the adaptation mechanism isn't necessary.  However, it is necessary for images that do not contain a shell, or that use the `ENTRYPOINT` feature of docker (in which case the environment cannot be overridden from the container's command without specifically overriding the entrypoint).  Personally, I also think it's arguable that the experience will be better to adapt services in this manner even when images have a shell and can perform the substitution themselves.

**Comment by smarterclayton** [comment]
Or even the mysql client (http://dev.mysql.com/doc/refman/5.0/en/environment-variables.html)

MYSQL_HOST
MYSQL_TCP_PORT

Neither of those match our existing env.

----- Original Message -----

**Comment by pmorie** [comment]
@smarterclayton and I discussed this offline and think the adaptation use-case doesn't depend on predeclaration of services nor does it need to be coupled to services at all.

Here's another use-case that might require pre-declaration:

Comment


1.  As a user I want to express that a pod should not be started unless services it depends on have had IP/Ports allocated

Consider the following to address that use-case:

    type PodSpec struct {
      // other fields omitted
      ServiceLinks []ServiceLinks
    }

    type ServiceLink struct {
      TypeMeta
      ObjectMeta

      Target    ObjectReference
      NeedReady bool
  }

`NeedReady` would be a precondition that states that the service must have at least one endpoint.

Any thoughts? @smarterclayton @erictune @bgrant0607 @thockin @lavalamp

**Comment by pmorie** [comment]
Also, in the context of the above, it would be good to produce an event after some time if a pod which is scheduled cannot start, but that's perhaps another issue.

**Comment by bgrant0607** [comment]
Previous discussion of this last type of dependency was in #2385. 

It has been my hope that DNS will eliminate the creation order problem between services and their clients. Creation-order dependencies are problematic since containers can go down at any time, and they have unclear meaning when updating or replacing objects (such as with rolling updates). 

That said, sometimes there are unavoidable turn-up dependencies, such as to initialize stateful services, such as databases or message brokers. I envision handling such dependencies in deployment automation: #1704.

**Comment by bgrant0607** [comment]
I just experienced this, as additional confirmation: The new directory-reading feature of kubectl is not so useful for containers using service environment variables. Someone tried to do:
```
cluster/kubectl.sh create -f examples/guestbook-go/
```

**Comment by smarterclayton** [comment]
Shame service is after controller in the alphabet

**Comment by pmorie** [comment]
@bgrant0607 @smarterclayton Suppose we avoid changing the pod model and just prototype expressing the dependency as an annotation on the pod?

**Comment by bgrant0607** [comment]
Note that the guestbook example use the convention of a `uses` label:
https://github.com/GoogleCloudPlatform/kubernetes/search?utf8=%E2%9C%93&q=uses+language%3Ajson+language%3Ayaml&type=Code

I'm not too keen on that approach. If an object name, an annotation would be better. Annotations could be understood by deployment tools, the CLI, and UIs, without being understood by the system, and could be an embedded ObjectReference.

Independently of creation order, we do need to either discontinue support for links entirely, or generate them only on demand.

**Comment by bgrant0607** [comment]
Sorry: @pmorie See my previous comment. For creation order, yes.

**Comment by pmorie** [comment]
@bgrant0607 I had noticed that in the guestbook and still wonder what the value-add of it is relative to that example.

**Comment by pmorie** [comment]
@bgrant0607 Am I correct that what you're thinking of regarding creation order is that tooling would ingest some set of resources to create, perform graph analysis on them to build a dependency graph, and order create calls such that a pod with dependencies on services X, Y would not be created until after services X and Y?

What about the other way around, where a pod could be created before its dependency services, but wouldn't be started by the kubelet until the services had been created?  In this case your pod may not start, but that's already the case - as an example, if a pod is scheduled onto a host and needs a port on that host which is already consumed, it won't start.

**Comment by pmorie** [comment]
@bgrant0607 Regarding this comment: 

> Independently of creation order, we do need to either discontinue support for links entirely, or generate them only on demand.

If we decide to retain links and make them explicit, do you think it would coexist with an annotation approach, or supercede it?  Ie, tooling could read the part of the model that expresses the link relationship and use that to infer the right creation order.

**Comment by pmorie** [comment]
@bgrant0607 let me add some detail to the above comment:

> What about the other way around, where a pod could be created before its dependency services, but wouldn't be started by the kubelet until the services had been created?  In this case your pod may not start, but that's already the case - as an example, if a pod is scheduled onto a host and needs a port on that host which is already consumed, it won't start.

There a couple use-cases that we've mixed in this issue, let me try to decompose them as follows:

1.  As a user, I want to declare that my pod has a dependency on some service X, and for my pod not to be started until service X exists and is **usable/ready**
2.  As a user, I want to declare that my pod has a dependency on some service X, and for my pod not to be started until service X exists
3.  As a user, I want to declare that my pod has a dependency on some service X, so that I can have some other tool that builds a dependency tree and figures out the right order to start in

For the sake of discussion, let's alias them respectively as follows:

1. "readiness before start"
2. "existence before start"
3. "external tooling"

Let's also keep in mind that there are at least three failure modes on the pod side:

1.  Container can't start because it does not have the coordinates to reach a dependency service
2.  Container can't start because a dependency service isn't ready after either attempting to retry and exhausting attempts or hitting a timeout
3.  Container can't start because a dependency service isn't ready, even though it is technically reachable (viz, pod has coordinates for the service, but gets a 503 because there are no endpoints), and the container doesn't know how to retry (ie, image developer does not have retries)

Failure mode 1 can be addressed with certainty; you can implement logic that doesn't attempt a pod start until its dependency services exist.

Failure mode 2 cannot be addressed with certainty, for the same reason as failure mode 3; you can't make real guarantees about readiness even with a health check.

Failure mode 3 cannot be addressed with certainty - even with a health check against dependency services, since you can make no guarantees about readiness, even with some kind of health check against a dependency service, since the service endpoints can all go down at any moment.

Given that only failure mode 1 can really be addressed with certainty, let's apply that to the above use cases we've been discussing:

1.  "readiness before start" does not have much utility.  Service endpoints can go away whenever, and there's no guarantee that a service that passed a health/readiness check will be healthy or ready when the pod starts
2.  "existence before start" solves failure mode 1; your pod might not ever start, but it might not start today, either
3.  "external tooling" potentially solves failure mode 1; your tooling will have to be able to accurately build and walk the dependency tree to ensure that it creates things in the right order

I think relying on things being created in the right order is flaky and imposes a lot on the tooling.  That doesn't mean we shouldn't consider it.  I do, think (maybe feel is more appropriate), though, that "existence before start" will meet the needs of ~80% of images.

I don't love the idea of adding polling loops in images.   @danmcp pointed out during a discussion we were having about this earlier that this would have negative influence on the effectiveness of throttling.

In case it's not clear, I think "existence before start" is the sweet spot to implement here.

cc @smarterclayton @danmcp

**Comment by bgrant0607** [comment]
First, creation-order dependencies:

@pmorie 

We have several forms of inter-object dependencies: label selectors, link variables, and, soon, direct object references, such as replication controller to pod template (#170). 

In at least some cases, we want to be able to give errors to users for genuine mistakes.

I would like kubectl and similar clients to be able to remain mostly relatively generic, without semantic knowledge of specific API types, other than common metadata, such as labels and annotations. Pulling the swagger or other reflection APIs would be acceptable -- for instance, I imagine we'll need something like that for reasoning about controllers and their controllees (see #3058). Dependency annotations specifically to drive creation workflow would be acceptable, also.

Here's a simple idea: Do the dependency checking on the apiserver. Return 412 Precondition Failed (anyone know of something more appropriate?) api.StatusReasonDependencyNotSatisfied if the dependency doesn't exist. kubectl can bump the object to the end of its queue and revisit it later, potentially being smart about it by comparing the reference cited in the error to the other items in its queue. If the required object isn't in the queue, or all the remaining items in the queue have failed, then it can stop trying and raise an error to the user.

**Comment by bgrant0607** [comment]
Now, service predeclaration:

I agree with @smarterclayton that we should allow the user to specify HOW the service appears in the pod: environment variable name(s), /etc/hosts, maybe NATed IP or port, ...

I'd put the list of ServiceLinks into Container rather than PodSpec, for finer-grain control. I don't imagine multiple containers in a pod needing to contact the same services, in general.

ServiceLink would be a subobject rather than a top-level object, so it wouldn't need TypeMeta or ObjectMeta.

Link variables currently look like:
```
REDIS_MASTER_SERVICE_HOST=10.0.0.11
REDIS_MASTER_SERVICE_PORT=6379
REDIS_MASTER_PORT=tcp://10.0.0.11:6379
REDIS_MASTER_PORT_6379_TCP=tcp://10.0.0.11:6379
REDIS_MASTER_PORT_6379_TCP_PROTO=tcp
REDIS_MASTER_PORT_6379_TCP_PORT=6379
REDIS_MASTER_PORT_6379_TCP_ADDR=10.0.0.11
```

For compatibility/convenience, we could have one option that creates all of those.

```
type ServiceLink struct {
	Target ObjectReference // service/portal
	Port util.IntOrString 

	DNS bool
	DockerLinks bool
	EtcHostsName string // create an entry in /etc/hosts with the specified name
	IPV4Var string // set the specified environment variable to the IPv4 address, if any
	IPV6Var string // set the specified environment variable to the IPv6 address, if any
	PortVar string // set the specified environment variable to the port, if any
	ProtocolVar string  // set the specified environment variable to the protocol
}
```

DNS for all services within the namespace would come for free, with DNS configured, but the DNS bool would declare the pre-existence dependency.

We should think about how to extend this to L7 routes and headless Endpoints. Maybe Target could be used in the L7 case, unless multiple Routes were folded into a single object, and for the Endpoints case, though in the latter case, we'd need a different mapping option (or none -- just use the API).

I'll also think about whether this could/should be generalized to other types of information for the downward API #386.

**Comment by smarterclayton** [comment]
I believe precondition 412 is reserved for actual http semantics.  422 with a reason is probably sufficient.

Do we need to distinguish between require dependencies and optional ones?  I *think* optional dependencies should be reflected in status.  Required dependencies can be described in the error, and causes can potentially list the error.  I think I buy the argument that for hard dependencies this is a good reqt.  Question is - when *should* hard requirements exist, and when shouldn't they?  Will circle with Paul more.

**Comment by bgrant0607** [comment]
422 - SGTM. Thanks. Makes sense.

I'd be fine with distinguishing optional vs. required.

**Comment by smarterclayton** [comment]
It we thought this was generalizable to downward api (for service like things), we might want to frame this like a formatter interface with simple env substitution.  The more ways to represent this info the harder it gets to add fields.  I don't think I've seen a "great" way to represent the idea.  There's some break even point probably where a user can do the medium hard thing (build a url that works is a good case).  Building urls is pretty common - for instance if you wanted to talk to an etcd cluster you'd need to know protocol, host, port, minimum.  But you might want the default service DNS name, or the NameRecord that Tim mentioned, or the public IP.  Most of that can be precalculated by the person who creates the pod, so more burden can reside there.

I'm starting to feel like adaptation and downward api would best be allowed in the short term with *only* the following:

* Introduce a Boolean value on api.EnvVar called "resolve", and then prior to execution do one pass to get base values, then do another pass and do simple substitution on any that need to be resolved.  Service vars would be in the first pass.
* allow "special" substitutions during resolve (downward API) if we have a use case - these values can only be used from a resolve, not a regular env.
* resolve command / Args / entrypoint (so someone can do more complex shell substitution predictably)

I think a resolution pass of this form gives a pod author more flexibility for a small increase in complexity.  Forcing a shell for complex expansion is not unreasonable, but this also forces someone to explicitly choose exposing a downward API into the pod, while preventing accidental expansion (opt in).

It would also allows us to simplify service link - we promise to add certain env, and they can always override that env on their own, or mutate it to fit their needs.

**Comment by pmorie** [comment]
@bgrant0607 I think your suggestion about the API server and `kubectl` is a good one.

@bgrant0607 @smarterclayton For OpenShift I think there is a case where we will require distinguishing between optional and required services.  The use case for optional services is one where you have some set of images which you want to play nicely with one another when they work together, but where independent usage is a valid use-case.  For example, consider `mysql` and `ruby` images that are designed to work together.  The `ruby` image should know how to consume a service for the `mysql` image, but you shouldn't have to use `mysql` in order to use `ruby`

**Comment by brendandburns** [comment]
I think I strongly believe that these kind of linkages should be the provence of things that live above the default Kubernetes layer, these things can have well understood semantics for linkages, but I'd really like to not have workflow anywhere in the core k8s API.

I like @smarterclayton 's small set of adaptations to the env var syntax.

wrt to labels vs. annotations, then I'm totally ok with using annotations (when they exist)

**Comment by pmorie** [comment]
@brendanburns Does your above comment mean that you don't think the core API should have the types necessary to express the relationships described here, even if they're dealt with solely at a higher level?

**Comment by bgrant0607** [comment]
To meet our 1.0 scaling goals, I'm pretty sure we'll have to stop adding all environment variables for all services to every pod by default. Pods will need to specify that they need them. We've talked about eliminating the variables entirely, but that would break O(all) images in the Docker registry.

**Comment by pmorie** [comment]
@bgrant0607 have you thought at all about what the API for this should be?

**Comment by thockin** [comment]
Now that we only add variables for the pod's own namespace, is this as big
a concern?  I see it as a short-to-medium term thing, but something that
~every pod will turn on, so why make it an option?

**Comment by pmorie** [comment]
@bgrant0607 Disregard this comment (https://github.com/GoogleCloudPlatform/kubernetes/issues/1768#issuecomment-76512182); I somehow missed your above comments.

**Comment by brendandburns** [comment]
I don't think this is relevant for 1.0.

Let's not over optimize ahead of time.  It's an easy, additive change of it
is required for scale (and I don't think it will be.)

Brendan

**Comment by brendandburns** [comment]
s/of/if/  stupid phone...

**Comment by brendandburns** [comment]
As above, I don't think this makes the v1.0 cut.

**Comment by smarterclayton** [comment]
Regarding disabling the default service env vars, I would like to specify a way for a pod to enable injection (have it default to true for now).  And then eventually remove it.  So either a generator or env option which is *bool and defaults to true for v1.

**Comment by smarterclayton** [comment]
Also actually allow you to generate their equivalent via downward API

**Comment by bgrant0607** [comment]
@smarterclayton I agree with adding an option that would allow us to turn off env vars.

For others: A more comprehensive linking proposal is in progress in #17543

**Comment by bgrant0607** [comment]
cc @ncdc

**Comment by bgrant0607** [comment]
I re-opened #7365. If we had a service downward API along the lines of what we have for ConfigMap and Secret, then Kubernetes could defer container start and/or pod creation until a required Service existed, could surface conditions and/or events about non-existent services, etc.

**Comment by zq-david-wang** [comment]
@smarterclayton  I notice a huge delay when spawning bash via "docker exec" if there are thousands of services (I tested with 3000) in the namespace. It would be great to have options to disable service env when creating docker process. 

The test I run is as following: (kubelet 1.6)
I have build a docker image with bash installed on alpine base image.
When run "docker exec -it [docker-id] bash", 
it took about 20s to run it if there are 3000 services within the namespace;
it took only less than 1s to run it if I disable the service env by modifying the code.

**Comment by bgrant0607** [comment]
cc @kubernetes/sig-network-feature-requests @kubernetes/sig-node-feature-requests

**Comment by endocrimes** [comment]
Now that service links are optional (dealing with some of the performance and collision issues that they used to cause), we're mostly waiting on a way to declare a subset of them for an application here, and it seems no headway has been made on that recently, I'm going to bump this down to important-longterm.

/priority important-longterm

**Comment by MadhavJivrajani** [comment]
/remove-kind design
/kind feature

`kind/design` will soon be removed from `k/k` in favor of `kind/feature`. Relevant discussion can be found here: https://github.com/kubernetes/community/issues/5641

**Comment by thockin** [comment]
Realistically, no.  Service links are not something we're going to do more to support.
